### PR TITLE
GEOPY-1688: allow baseDriver to accept BaseData

### PR DESCRIPTION
**Pull request by MatthieuCMira**
**GEOPY-1688 - allow baseDriver to accept BaseData**
Accept BaseData as params

### CODE REVIEW

File: geoapps_utils/driver/driver.py [hunk 1]
```diff
@@ -25,7 +26,7 @@ class BaseDriver(ABC):
     _params_class = BaseParams
     _validations: dict | None = None
 
-    def __init__(self, params: BaseParams):
+    def __init__(self, params: BaseParams | BaseData):
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:29]: While at it, can we make the input attributes more explicit in the init?

```suggestion
    def __init__(self, params: BaseParams | BaseData):
        self.params = params
        self.workspace = params.geoh5
        self.out_group = params.out_group
```

File: geoapps_utils/driver/driver.py [hunk 2]
```diff
@@ -25,7 +26,7 @@ class BaseDriver(ABC):
...
         """
         :param params: Application parameters.
         """
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:32]: This goes in the signature of the class.

**Review by domfournier — COMMENTED** [review]
looks good. Minor comments. Also we might have to remove this line. If the pydantic class is frozen, then this would make it to crash

image

**Comment by MatthieuCMira** [comment]
Overall question:

I removed some __init__ from the assets. 
Pylint did not like this file, and I had no idea why it was there. 
Is there a good reason to keep this file?

**Comment by MatthieuCMira** [comment]
Cannot change much in BaseDriver.
Simpeg Driver seems quite hacky using not straight forwards mechanics...

File: geoapps_utils/driver/data.py [hunk 3]
```diff
@@ -133,7 +133,10 @@ def input_file(self) -> InputFile:
             ifile = copy(self._input_file)
             ifile.validate = False
 
-        ifile.data = dict(ifile.data, **self.flatten())
+        self_params = {**self.flatten()}
+        ifile.data = {
+            key: self_params.get(key, value) for key, value in ifile.data.items()
+        }
```
Comment (domfournier) [inline geoapps_utils/driver/data.py:139]: Why the change? Both seems to do the same thing, just more compact.

**Comment by MatthieuCMira** [comment]
> looks good. Minor comments. Also we might have to remove this line. If the pydantic class is frozen, then this would make it to crash
> 
> image

This line is probably the result of some hacky Simpeg driver processes. 
We should first refactor this package before changing it

**Comment by MatthieuCMira** [comment]
I added a function to create the params in the driver directly from a uijson. 

So driver creation can be more straight forward 

(option for BaseData Only)

File: geoapps_utils/driver/driver.py [hunk 4]
```diff
@@ -38,6 +38,34 @@ def __init__(self, params: BaseParams | BaseData):
...
+
+        # verify params_dict is baseData to run this line
+        if (
+            isinstance(input_file, InputFile)
+            and input_file.data is not None
+            and issubclass(self._params_class, BaseData)
+        ):
+            params_dict = input_file.data
+            params_dict["_input_file"] = input_file
+            return self._params_class(**params_dict)
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:62]: Use the `build` method of `BaseData` instead so we can handle nested Models

File: geoapps_utils/driver/driver.py [hunk 5]
```diff
@@ -38,6 +38,34 @@ def __init__(self, params: BaseParams | BaseData):
...
+
+        raise TypeError(
+            "Input file must be of type str, Path, or geoh5py.ui_json.InputFile,"
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:65]: Split the raise so it's clearer why it failed and do it before the return. First check that input_file was valid, then check that `_param_class` is type BaseData.

File: geoapps_utils/driver/driver.py [hunk 6]
```diff
@@ -38,6 +38,34 @@ def __init__(self, params: BaseParams | BaseData):
         if hasattr(self.params, "out_group") and self.params.out_group is None:
             self.params.out_group = self.out_group
 
+    def create_from_input_file(self, input_file: str | Path | InputFile):
+        """
+        Create driver from input file.
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:43]: Looks like it's creating a params, not a driver. Type the return too...

File: geoapps_utils/driver/driver.py [hunk 7]
```diff
@@ -50,6 +78,9 @@ def params(self):
 
     @params.setter
     def params(self, val: BaseParams | BaseData):
+        if isinstance(val, (str, Path, InputFile)):
```
Comment (domfournier) [inline geoapps_utils/driver/driver.py:81]: I am not a fan of this, we are mixing what params can be. Feels like you want to have an `input_file` setter instead, and do the setting there instead.

**Review by domfournier — COMMENTED** [review]
I am not sold on the `create_from_input_file` method. It is duplicating a lot of the mechanics of `BaseData.build`, and mixing the input type for params...

**Comment by MatthieuCMira** [comment]
Well, the "start" classmethod is already waht we wanted to do.

**Review by benk-mira — CHANGES_REQUESTED** [review]
I think we should spend a bit more time on the input_file stuff.  It seems like we're losing the ability to carry extra parameters through.

**Review by benk-mira — CHANGES_REQUESTED** [review]

**Review by benk-mira — APPROVED** [review]

**Pull request merged**